daemon/start.go

```diff
@@ -197,4 +198,10 @@ func (daemon *Daemon) Cleanup(container *container.Container) {
 		}
 	}
 	container.CancelAttachContext()
+
+	// if containers AutoRemove flag is set, remove it after clean up
+	if container.HostConfig.AutoRemove {
+		// If containers lock is not released, goroutine will guarantee no block
+		go daemon.ContainerRm(container.ID, &types.ContainerRmConfig{ForceRemove: true, RemoveVolume: true})
```

Comment:
I hope I can, then I don't need to do so much works to guarantee the client must exit after container is auto-removed, but actually it's impossible. 

If I'm not doing it in a goroutine, it will always block in current design. 
For example:  https://github.com/docker/docker/blob/master/daemon/monitor.go#L42, https://github.com/docker/docker/blob/master/daemon/start.go#L114.
This `Cleanup` function is always invoked in locked state, also `daemon.ContainerRm` need to lock/unlock container from time to time, so it will *always* block.

Using goroutine is the most convenient way, and it doesn't hurt in my opinion.